python/cuml/internals/global_settings.py

```diff
@@ -97,3 +121,42 @@ def output_type(self):
     @output_type.setter
     def output_type(self, value):
         self._output_type = value
+
+    @property
+    def xpy(self):
+        return self.memory_type.xpy
+
+
+global_settings = GlobalSettings()
+
+
+def set_global_memory_type(memory_type):
+    global_settings.memory_type = MemoryType.from_str(memory_type)
+
+
+class using_memory_type:
+    def __init__(self, memory_type):
+        self.prev_memory_type = global_settings.memory_type
+        set_global_memory_type(memory_type)
+
+    def __enter__(self):
+        return self.prev_memory_type
+
+    def __exit__(self, type_, value, traceback):
+        set_global_memory_type(self.prev_memory_type)
```

Comment:
Yep, but we want to avoid `contextlib.contextmanager` wherever possible until/unless its performance problems are fixed. I've pinged you offline with more details on this one.